plugin/pkg/admission/security/podsecuritypolicy/admission.go

```diff
@@ -123,7 +123,7 @@ func (c *PodSecurityPolicyPlugin) Admit(a admission.Attributes) error {
 	pod := a.GetObject().(*api.Pod)
 
 	// compute the context
-	allowedPod, pspName, validationErrs, err := c.computeSecurityContext(a, pod, true)
+	allowedPod, pspName, validationErrs, err := c.computeSecurityContext(a, pod, true, "")
```

Comment:
When creating pods, if a user sets the kubernetes.io/psp annotations, we can also set this psp the most preferred one.
But I am not sure whether this is good.

Reply:
no, let's only do this on Validate. Pushing one to the front for mutation cases can change which one takes priority, which isn't good. During Validation, no mutation is allowed, so priority doesn't really matter

Reply:
I don't like this either. Following the annotation gives the false impression that the given psp is mandatory to choose. But it won't be used as soon as it does not match.

Reply:
Can you add a comment that this  `""` is intentional, i.e. that the annotation is not taken into account?

Having said that: shouldn't we make a difference between CREATE and UPDATE? On create we want to ignore the annotation. But on update we could use it to gain the same speed-up in the mutating admission phase.

@liggitt wdyt?

Reply:
> But on update we could use it to gain the same speed-up in the mutating admission phase.

On UPDATE this function will not run.
ref: [line 119](https://github.com/kubernetes/kubernetes/pull/56843/files#diff-291b8dd7d08cc034975ddb3925dbb08fR119)

Reply:
Done.

Reply:
> On UPDATE this function will not run.

Ok, then my memory was wrong that we hadn't optimized that yet.

Reply:
On CREATE verb, `Admit()` and `Validate()` will run. `computeSecurityContext()` will run twice. The second time is optimized.

On UPDATE verb, `Validate()` will will run. `computeSecurityContext()` will run once. It is optimized.

Reply:
:+1: